qhdopt/backend/dwave_backend_sim.py

```diff
@@ -0,0 +1,150 @@
+from typing import Tuple, List
+
+from simuq import QSystem, Qubit
+from simuq.dwave import DWaveSimProvider
+import numpy as np
+import time
+from qhdopt.utils.decoding_utils import spin_to_bitstring
+
+from qhdopt.backend.backend import Backend
+
+
+class DWaveBackendSim(Backend):
+    """
+    Backend implementation for D-Wave simulated annealing using neal.
+    This backend doesn't require an API key and runs simulated annealing locally.
+    """
+    def __init__(self,
+                 resolution,
+                 dimension,
+                 univariate_dict,
+                 bivariate_dict,
+                 shots=100,
+                 embedding_scheme="unary",
+                 anneal_schedule=None,
+                 penalty_coefficient=0,
+                 penalty_ratio=0.75,
+                 chain_strength_ratio=1.05,
+                 **sampler_kwargs):
+        super().__init__(resolution, dimension, shots, embedding_scheme, univariate_dict,
+                         bivariate_dict)
+        if anneal_schedule is None:
+            anneal_schedule = [[0, 0], [20, 1]]
+        self.anneal_schedule = anneal_schedule
+        self.penalty_coefficient = penalty_coefficient
+        self.penalty_ratio = penalty_ratio
+        self.chain_strength_ratio = chain_strength_ratio
+        # Store additional kwargs for SimulatedAnnealingSampler
+        self.sampler_kwargs = sampler_kwargs
+
+    def calc_penalty_coefficient_and_chain_strength(self) -> Tuple[float, float]:
+        """
+        Calculates the penalty coefficient and chain strength using self.penalty_ratio.
+        """
+        if self.penalty_coefficient != 0:
+            chain_strength = np.max([5e-2, self.chain_strength_ratio * self.penalty_coefficient])
+            return self.penalty_coefficient, chain_strength
+          
+        qs = QSystem()
+        qubits = [Qubit(qs) for _ in range(len(self.qubits))]
+        qs.add_evolution(self.S_x(qubits) + self.H_p(qubits, self.univariate_dict, self.bivariate_dict), 1)
+        dwp = DWaveSimProvider(**self.sampler_kwargs)
+        h, J = dwp.compile(qs, self.anneal_schedule)
+        max_strength = np.max(np.abs(list(h) + list(J.values())))
+        penalty_coefficient = (
+            self.penalty_ratio * max_strength if self.embedding_scheme == "unary" else 0
+        )
+        # For simulated annealing, chain_strength is not as critical but we keep it for consistency
```

Comment:
It's actually not necessary, so we'd rather remove it instead of cluttering the code